lib/utils/kalman_higher_derivatives.cpp

```diff
@@ -0,0 +1,59 @@
+#include "kalman_higher_derivatives.hpp"
+
+#include <algorithm>
+
+namespace hyped::utils {
+KalmanHigherDerivatives::KalmanHigherDerivatives(core::ILogger &logger,
+                                                 const core::ITimeSource &time)
+    : logger_(logger),
+      time_(time),
+      sum_jerk_(0),
+      sum_snap_(0)
+{
+  acceleration_values_.resize(kNumDataPoints);
+  std::fill(acceleration_values_.begin(), acceleration_values_.end(), 0.0F);
+
+  timestamps_.resize(kNumDataPoints);
+  std::fill(timestamps_.begin(), timestamps_.end(), 0);
+
+  jerk_values_.resize(kNumDataPoints - 1);
+  std::fill(jerk_values_.begin(), jerk_values_.end(), 0.0F);
+
+  snap_values_.resize(kNumDataPoints - 2);
+  std::fill(snap_values_.begin(), snap_values_.end(), 0.0F);
+}
+
+navigation::HigherDerivatives KalmanHigherDerivatives::getHigherDerivatives(
+  const std::uint64_t timestamp, const core::Float acceleration)
+{
+  // update acceleration and timestamps vectors
+  acceleration_values_.pop_back();
+  acceleration_values_.insert(acceleration_values_.begin(), acceleration);
+
+  timestamps_.pop_back();
+  timestamps_.insert(timestamps_.begin(), timestamp);
+
+  // calculate new jerk value from new data and update jerk sum
+  sum_jerk_ -= jerk_values_.at(jerk_values_.size() + 1);
```

Comment:
probably stupid but not too sure why we do this?

Reply:
Instead of summing the whole array each time the function is called (which is $O(n)$), we store the total of the array as a data class member, subtract the element at the back of the array (that we are removing) and add the new element that we're inserting at the front of the array to the total (constant time).